### ISSUE TITLE
Replication freezes if slave bumps LSN while master is restarted

### ISSUE THREAD

**Issue by locker** [tarantool/tarantool#3038]
Branch 1.7-next, commit ad237aebef1a3b605413bca8bc31ebf362db9e90

1. Setup the master:
```box.cfg{listen=12345}```
```box.schema.user.grant('guest', 'replication')```
```box.schema.space.create('test')```
```box.space.test:create_index('pk')```
2. Setup the slave:
```box.cfg{replication=12345}```
3. Temporarily shutdown the master.
4. Bump LSN on the slave.
```box.space.test:delete{1}```
5. Restart the master.

Now, the slave reports that it follows the master, but in fact it doesn't:

- Master
```
tarantool> for i = 1, 100 do box.space.test:auto_increment{} end
---
...

tarantool> box.info
---
- version: 1.7.6-143-gad237aeb
  id: 1
  ro: false
  vinyl: []
  vclock: {1: 105}
  uptime: 250
  lsn: 105
  memory: []
  cluster:
    uuid: 2a96261e-cc0c-4800-999c-f8c285e047fa
  pid: 29378
  status: running
  signature: 105
  replication:
    1:
      id: 1
      uuid: 5b415f9a-8778-450b-8b98-44c99f96ae85
      lsn: 105
    2:
      id: 2
      uuid: 9fc38b0e-b743-4347-befd-8572158d7274
      lsn: 0
      downstream:
        vclock: {2: 1, 1: 5}
  uuid: 5b415f9a-8778-450b-8b98-44c99f96ae85
...
```
- Slave
```
tarantool> box.space.test:count()
---
- 0
...

tarantool> box.info
---
- version: 1.7.6-143-gad237aeb
  id: 2
  ro: false
  vinyl: []
  vclock: {2: 1, 1: 5}
  uptime: 344
  lsn: 1
  memory: []
  cluster:
    uuid: 2a96261e-cc0c-4800-999c-f8c285e047fa
  pid: 29361
  status: running
  signature: 6
  replication:
    1:
      id: 1
      uuid: 5b415f9a-8778-450b-8b98-44c99f96ae85
      lsn: 5
      upstream:
        status: follow
        idle: 0.91151907200401
        peer:
        lag: 8.1539154052734e-05
    2:
      id: 2
      uuid: 9fc38b0e-b743-4347-befd-8572158d7274
      lsn: 1
  uuid: 9fc38b0e-b743-4347-befd-8572158d7274
...
```
After calling `box.snapshot()` on the master, replication stops with an error:
```
tarantool> box.snapshot()
2018-01-11 19:29:37.124 [29378] snapshot/101/main I> saving snapshot `./00000000000000000105.snap.inprogress'
2018-01-11 19:29:37.154 [29378] snapshot/101/main I> done
---
- ok
...

tarantool> for i = 1, 100 do box.space.test:auto_increment{} end
2018-01-11 19:29:41.773 [29378] relay/127.0.0.1:34878/101/main recovery.cc:295 E> XlogGapError: Missing .xlog file between LSN 6 {1: 5, 2: 1} and 105 {1: 105}
2018-01-11 19:29:41.774 [29378] relay/127.0.0.1:34878/101/main C> exiting the relay loop
---
...
```

**Comment by locker** [comment]
This bug seems to be the real cause of replication failures described in #2825 and #2940

**Comment by locker** [comment]
Fixed by e755ad247c0f8b9ec0c05fdf4091b738864f8ff1
